habitat/tasks/nav/nav.py

```diff
@@ -514,6 +521,118 @@ def get_observation(
+            cat_mapping = MP3D_CATEGORY_TO_TASK_CATEGORY_ID
+        else:
+            cat_mapping = HM3D_CATEGORY_TO_TASK_CATEGORY_ID
+        self.category_to_task_category_id = cat_mapping
+        if config.RAW_NAME_TO_CATEGORY_MAPPING != "":
+            with open(config.RAW_NAME_TO_CATEGORY_MAPPING, "r") as fp:
+                lines = fp.readlines()
+            lines = lines[1:]
+            lines = [l.strip().split("    ") for l in lines]
+            self.raw_to_cat_mapping = {}
+            for l in lines:
+                raw_name = l[1]
+                cat_name = l[-1]
+                if cat_name in cat_mapping:
+                    self.raw_to_cat_mapping[raw_name] = cat_name
+        else:
+            self.raw_to_cat_mapping = {k: k for k in cat_mapping.keys()}
+
+    def _get_sensor_type(self, *args: Any, **kwargs: Any):
+        return SensorTypes.COLOR
+
+    def _get_observation_space(self, *args: Any, **kwargs: Any):
+        if self.config.CONVERT_TO_RGB:
+            observation_space = spaces.Box(
+                low=0,
+                high=255,
+                shape=(self.config.HEIGHT, self.config.WIDTH, 3),
+                dtype=np.uint8,
+            )
+        else:
+            observation_space = spaces.Box(
+                low=np.iinfo(np.int32).min,
+                high=np.iinfo(np.int32).max,
+                shape=(self.config.HEIGHT, self.config.WIDTH),
+                dtype=np.int32,
+            )
+        return observation_space
+
+    def get_observation(
+        self, *args: Any, observations, episode, **kwargs: Any
+    ):
+        episode_uniq_id = f"{episode.scene_id} {episode.episode_id}"
+        if self._current_episode_id != episode_uniq_id:
+            self._current_episode_id = episode_uniq_id
+            # Get mapping from instance id to task id
+            scene = self._sim.semantic_annotations()
+            self.instance_id_to_task_id = np.ones(
+                (len(scene.objects), ), dtype=np.int64
+            ) * -1 # Non-task objects are set to -1
+            for obj in scene.objects:
+                if obj is None:
+                    continue
+                obj_inst_id = int(obj.id.split("_")[-1])
+                obj_name = obj.category.name()
+                if obj_name in self.raw_to_cat_mapping:
+                    obj_name = self.raw_to_cat_mapping[obj_name]
+                    obj_task_id = self.category_to_task_category_id[obj_name]
+                    self.instance_id_to_task_id[obj_inst_id] = obj_task_id
+        # Set invalid instance IDs to unknown object 0
+        semantic = np.copy(observations["semantic"])
+        semantic[semantic >= self.instance_id_to_task_id.shape[0]] = 0
+        # Map from instance id to task id
+        semantic_category = np.take(self.instance_id_to_task_id, semantic)
+        if self.config.CONVERT_TO_RGB:
+            semantic_category = self.convert_semantic_to_rgb(semantic_category)
```

Comment:
RGB conversion should live in `def observations_to_image(observation: Dict, info: Dict) -> np.ndarray:`.
Something like this:
```
    if "semantic_category" in observation:
        flat_sem = observation[
            obj_semantic_name
        ]  # to move to same scale #.permute(2, 0, 1).unsqueeze(0).data.max(1)[1].cpu().numpy()[0]
        flat_sem[flat_sem == 41] = 40
        if not isinstance(flat_sem, np.ndarray):
            flat_sem = flat_sem.cpu().numpy()
        semantic_segmentation = (
            color_label(flat_sem).squeeze().transpose(1, 2, 0).astype(np.uint8)
        )
        egocentric_view.append(semantic_segmentation)

        if "objectgoal" in observation and "episode_info" in info:
            from habitat.tasks.nav.object_nav_task import task_cat2mpcat40

            # permute tensor to dimension [CHANNEL x HEIGHT X WIDTH]
            idx = task_cat2mpcat40[observation["objectgoal"][0]]
            goal_segmentation = (
                color_label(flat_sem == idx)
                .squeeze()
                .transpose(1, 2, 0)
                .astype(np.uint8)
            )
            egocentric_view.append(goal_segmentation)
```

Reply:
This sensor is not for visualization purposes. It is intended to be a model input for training with oracle semantics. So the logic should live within the sensor itself, right?